Marlin/src/HAL/STM32/inc/Conditionals_adv.h

```diff
@@ -24,3 +24,9 @@
 #if defined(USBD_USE_CDC_MSC) && DISABLED(NO_SD_HOST_DRIVE)
   #define HAS_SD_HOST_DRIVE 1
 #endif
+
+// Fix F_CPU not being a compile-time constant in STSTM32 framework
+#ifdef BOARD_F_CPU
+  #undef F_CPU
+  #define F_CPU BOARD_F_CPU
+#endif
```

Comment:
I don't know if this matters, but this hack will cause any external headers that use `F_CPU` and follow the include of `MarlinConfigPre.h` to have a different value of `F_CPU` than their corresponding `.cpp` files. This could have hard-to-diagnose side effects in current or future external libraries used by Marlin.

Reply:
That's exactly a side effect I'm looking for. External libraries will be redirected to a compile time version of the frequency, instead of a runtime version. That's very important to reach the last few level of optimization.

Reply:
External libraries won't see the `F_CPU` value set here. They will only see the `F_CPU` value set in the headers that they include. Do we expect all external libraries to use `BOARD_F_CPU` instead of `F_CPU` when `BOARD_F_CPU` is defined?

Reply:
Actually, one of the most painful external library is NeoPixel and it includes Delay.h from Marlin. This, in turns, includes Marlin_pre_config.h which will swap F_CPU for the compile time version. I would expect that, either the library does not make use of very small delay and in that case, using either F_CPU will work. Either the library requires small delay, and it'll likely need the declaration of DELAY_NS (via Delay.h) and in that case, F_CPU will be swapped to the compile time version and the DELAY_NS macro to convert to cycles will be compile time computed.

Reply:
> Actually, one of the most painful external library is NeoPixel and it includes `Delay.h` from Marlin.

When did that happen? If a library really needs some custom code to be optimally compiled inline without requiring any function calls, it should be provided as a template argument during instantiation. If the NeoPixel library is checking for `__MARLIN_FIRMWARE__` and including code from Marlin, that's quite a brittle dependency.

Reply:
Like [this](https://github.com/ccccmagicboy/Adafruit_NeoPixel/blob/087d9f714e104426029b7cbbddfd8bae7371710d/Adafruit_NeoPixel.cpp#L51) or [this](https://github.com/ellensp/Adafruit_NeoPixel/blob/093cd8bb94aef95b25166a3f79becc2b5c9f4e09/Adafruit_NeoPixel.cpp#L52). I agree it's not clean at all, but it's the only viable solution for STM32 right now. I'm preparing another large PR to remove this dependency and instead move the NeoPixel specific code into Marlin's respective HAL so that:
1. We can remove Adafruit NeoPixel dependency (and probably move the fix to the common build flags to the right place)
2. Implement the right delay/pin control by not burning CPU cycles with interrupt disabled (like it's currently doing). Each microcontroller has its own way to perform this. I would say that the default implementation will simply be "burn CPU time" but advanced implementation (using STM32 timers and DMA) can be written to optimize this whenever possible.
3. Allow changing the pixels value at any time and not only between layers since the (optimized) implementation could run with interrupt enabled

In all case, this PR is required to reach very small delay on STSTM32 framework, even if not talking about Neopixels.

Reply:
BTW, I'm not showing the current git of Adafruit_Neopixels because, IMHO, it should not be used, as it's terribly dangerous to use on STM32. It's re-programming the SYSCLK timer to simulate neopixel's timing, but this will completely break Marlin internals.

Reply:
I think that was already discussed somewhere. @sjasonsmith may remember about it.

Reply:
Is that [this](https://github.com/MarlinFirmware/Marlin/pull/19059#issuecomment-675660640) ?

Reply:
But that is in SAMD51 framework...

Reply:
I agree the best idea is to just include our own NeoPixel driver. Let's make it so!

Reply:
P.S. I think this line can probably go into `pins_postprocess.h` since it pertains to a setting expected to come from one of the pins files (if not a compile-time define).